factor/telegram/classes/factor.php

```diff
@@ -0,0 +1,315 @@
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
+        $record = $DB->get_record_sql($sql, array($USER->id, 'telegram'));
+
+        if ($enteredcode == $record->secret) {
+            if ($record->timecreated + $duration > time()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Cleans up email records once MFA passed.
+     *
+     * {@inheritDoc}
+     */
+    public function post_pass_state() {
+        global $DB, $USER;
+        // Delete all telegram records except base record.
+        $selectsql = 'userid = ?
+                  AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
+        $DB->delete_records_select('tool_mfa', $selectsql, array($USER->id, 'telegram'));
+
+        // Update factor timeverified.
+        parent::post_pass_state();
+    }
+
+    /**
+     * TOTP Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_factor_form_definition($mform) {
+        $mform->addElement('text', 'telegramuserid', get_string('telegram:telegramuserid', 'factor_telegram'));
+        $mform->setType('telegramuserid', PARAM_ALPHANUM);
+
+        return $mform;
+    }
+
+    /**
+     * TOTP Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        global $DB, $USER;
+
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = \'telegram\'
+                   AND label LIKE \'telegram:%\'';
+        $record = $DB->get_record_sql($sql, array($USER->id));
+
+        if (!empty($data->telegramuserid)) {
+            $row = new \stdClass();
+            $row->userid = $USER->id;
+            $row->factor = $this->name;
+            $row->label = 'telegram:'.$data->telegramuserid;
+            $row->timecreated = time();
+            $row->createdfromip = $USER->lastip;
+            $row->timemodified = time();
+            $row->lastverified = time();
+            $row->revoked = 0;
+
+            if (empty($record)) {
+                $id = $DB->insert_record('tool_mfa', $row);
+            } else {
+                $id = $row->id = $record->id;
+                $DB->update_record('tool_mfa', $row);
+            }
+
+            $record = $DB->get_record('tool_mfa', array('id' => $id));
+            $this->create_event_after_factor_setup($USER);
+
+            return $record;
+        }
+
+        return null;
+    }
+
+    /**
+     * Email factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function possible_states($user) {
+        // Email can return all states.
+        return array(
+            \tool_mfa\plugininfo\factor::STATE_FAIL,
```

Comment:
I can't see any code handling an explicit FAIL so I think this is best removed for the MVP